tests/testthat/test-linear_pool.R

```diff
@@ -453,13 +453,130 @@ test_that("samples only collected and re-indexed for simplest case", {
+      comp_unit_cols = c("target", "location"),
       n_output_samples = NULL
     )
 
   expect_equal(actual_outputs, expected_outputs)
 })
 
 
+test_that("ensemble of samples correctly drawn for compound task ID sets", {
+  sample_outputs <- expand.grid(stringsAsFactors = FALSE,
+                                model_id = letters[1:4],
+                                location = c("222", "888"),
+                                horizon = 1, #week
+                                target = "inc death",
+                                target_date = as.Date("2021-12-25"),
+                                output_type = "sample",
+                                output_type_id = 1:3,
+                                value = NA_real_)
+
+  sample_outputs$value[sample_outputs$location == "222" &
+                         sample_outputs$output_type_id == 1] <-
+    c(40, 30, 45, 80)
+  sample_outputs$value[sample_outputs$location == "222" &
+                         sample_outputs$output_type_id == 2] <-
+    c(60, 40, 75, 20)
+  sample_outputs$value[sample_outputs$location == "222" &
+                         sample_outputs$output_type_id == 3] <-
+    c(10, 70, 15, 50)
+  sample_outputs$value[sample_outputs$location == "888" &
+                         sample_outputs$output_type_id == 1] <-
+    c(100, 325, 400, 300)
+  sample_outputs$value[sample_outputs$location == "888" &
+                         sample_outputs$output_type_id == 2] <-
+    c(250, 350, 500, 250)
+  sample_outputs$value[sample_outputs$location == "888" &
+                         sample_outputs$output_type_id == 3] <-
+    c(150, 300, 500, 350)
+  sample_outputs <- sample_outputs |>
+    dplyr::mutate(horizon = 0, target_date = target_date - 7) |>
+    dplyr::bind_rows(sample_outputs)
+
+  set.seed(1234)
+  models_to_resample <- sample(x = letters[1:4], size = 5 %% 4)
+
+  # All compound units have unique ids which are shared across dependent task columns
+  # expected model is re-sampled
+  subset_expected <- expand.grid(
+    stringsAsFactors = FALSE,
+    KEEP.OUT.ATTRS = FALSE,
+    target = "inc death",
+    location = c("222", "888"),
+    component_model = letters[1:5],
+    distinct_ids = 2
+  ) |>
+    dplyr::mutate(component_model = ifelse(component_model == letters[5], models_to_resample, component_model)) |>
+    dplyr::arrange(dplyr::across(c("target", "location"))) |>
+    dplyr::tibble()
+  subset_actual <- sample_outputs |>
+    dplyr::mutate(component_model = .data[["model_id"]], .before = 1) |>
+    linear_pool(
+      task_id_cols = c("target_date", "target", "horizon", "location"),
+      comp_unit_cols = c("target", "location"),
+      n_output_samples = 5
+    ) |>
+    dplyr::group_by(dplyr::across(dplyr::all_of(c("target", "location", "output_type_id", "component_model")))) |>
+    dplyr::summarize(distinct_ids = dplyr::n()) |>
+    dplyr::ungroup() |>
+    dplyr::select(-"output_type_id")
+  expect_equal(subset_actual, subset_expected)
```

Comment:
Also, do we check that output type ids from different component models are distinct in the ensemble output?  I see you are doing something with a variable called `distinct_ids` here, but I'm not totally clear on what that variable is doing.

Reply:
Overall, it would be helpful to me to enumerate (in a comment) the different things we expect to see in the output from this function and how we're checking it.

Reply:
Basically, this test is checking that for requested `n_output_samples` that don't divide evenly among models, the remainder is correctly split among the models (i.e. all the remainder samples aren't requested from one single model). I probably should actually do `n_output_samples = 6` to better test this code.

The `distinct_ids` is used to show if the dependencies are properly being accounted for, i.e. for this example in the test the same forecast (generated by output type id) is being drawn for horizons 0 and 1 for the same compound task set values. I don't have an explicit test that the output type ids from different component models are distinct in the ensemble output but I will add one.

Reply:
restating the second paragraph above: within each group defined by values of model id and comp_taskid_set variables, each output_type_id shows up for all values of variables that are not in the comp_taskid_set (i.e., for all horizons if we get trajectories)

Reply:
Would it be useful to add an in-line comment within the testing file to explain this?

Reply:
Yeah, I think that would be good! Thanks.

Reply:
I'm marking this comment as resolved.  there are a couple of things in here that are repeated in other comments.